ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -105,10 +105,14 @@ data NoSuchHonestChainSchema =
 genHonestRecipe :: QC.Gen HonestRecipe
 genHonestRecipe = sized1 $ \sz -> do
     (Kcp k, Scg s, Delta d) <- genKSD
-    -- 2s slots has at least 2k blocks. But that is ensuring k-1 more blocks
+    -- 2s slots have at least 2k blocks. But that is ensuring k-1 more blocks
     -- than we actually need. Thus it's safe to remove the k-1 last slots.
-    -- Therefore we set for a length of at least 2s - (k - 1).
-    l <- (+ (2*s - (k - 1))) <$> QC.choose (0, 5 * sz)
+    -- Therefore we first set for a length of at least 2s - (k - 1).
+    --
+    -- If we then add k+d slots (to get a total size of 2s+d+1), we get enough
```

Comment:
Heavily edited my previous comment.

Reply:
> At the moment Kcp is always chosen as 2 or more.

If we're relying on that, we should enforce it in `checkHonestRecipe`.

Reply:
How about the following?

```
{- |

Note [Minimum schema length]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Consider this diagram of a necessary prefix of the schemas' slots (the schemas
may have a suffix of more slots, not shown).

@
    GA................B......C.......D
@

* @G@ is the genesis point.

* @A@ is the first slot in the schema.

* @B@ is the @2s - (k - 1)@th slot in the schema.

* @C = B + d@.

* @D = C + k@.

Let sai abbreviate /slot after the intersection/. Let asai abbreviate /active
sai/.

Suppose @G@ is the intersection, and consider these facts.

* The hypothetical closed interval @(A, 2s)@ (which might exceed @D@) would
  contain at least @2k@ honest asais. Thus the interval @(A, B)@ contains at
  least @2k-k+1=k+1@ honest asais; so @k+1@ honest asais are ensured.

* All slots @>C@ can be unstable, since both requirements are satisfiable.
  First, @C@ is at least @d@ after the @k+1@st honest asai (which is @<=B@), so
  @C+1@ can be the @k+1@st adversarial asai, if necessary. Second, the
  generators require @1<k@ so that @A@ can be adversarial active if necessary,
  which ensures that the slot @C+1@ is more than @s@ after the first
  adversarial asai.

* Since they are unstable, the @k@ slots in the half-open interval @[C, D)@ can
  be adversarial active. Together with @A@ if necessary, this ensures at least
  @k+1@ adversarial asais.

* Therefore, @2s + 2k + d - 1@ ensures @G@ is a valid intersection.

-}
```

Reply:
> we should enforce it in checkHonestRecipe

Done in 9b0a6ee.

> How about the following?

This would be a bit harder to integrate. But I adopted a simpler diagram in 43ac3c9f and made a pass to clarify and relate the explanations.

The main difficulty I find with just switching to your version verbatim is that it starts with a few definitions and arguments without conveying the purpose of them other than the title.

Since I don't have much hints on what makes it better than the current explanation, I'm also limited to iterate effectively on the text. Happy to discuss it further, though, if you think it is worth the effort.